### PR TITLE
test: migrate build_script_env to snapbox

### PR DESCRIPTION
**Pull request by heisen-li**
### What does this PR try to resolve?

part of https://github.com/rust-lang/cargo/issues/14039.

### CODE REVIEW

File: tests/testsuite/build_script_env.rs [hunk 1]
```diff
@@ -171,17 +186,28 @@ fn rustc_bootstrap() {
...
+[ERROR] Cannot set `RUSTC_BOOTSTRAP=1` from build script of `foo v0.0.1 ([ROOT]/foo)`.
+[NOTE] Crates cannot set `RUSTC_BOOTSTRAP` themselves, as doing so would subvert the stability guarantees of Rust for your project.
+[HELP] If you're sure you want to do this in your project, set the environment variable `RUSTC_BOOTSTRAP=1` before running cargo instead.
+
+"#]])
         .with_status(101)
         .run();
 }
 
+#[allow(deprecated)]
```
Comment (heisen-li) [inline tests/testsuite/build_script_env.rs:210]: If the following test cases are all expanded, the content will be very large. I don't think of a better solution, maybe the current treatment is the most appropriate.

Comment (epage) [inline tests/testsuite/build_script_env.rs:210]: Highlighting this for better tracking

File: tests/testsuite/build_script_env.rs [hunk 2]
```diff
@@ -237,6 +264,7 @@ fn build_script_sees_cfg_target_feature() {
 /// what the value of `RUSTFLAGS` should be in this case. We chose to give a
 /// warning. However, no matter what we do, it's important that build scripts
 /// and rustc see a consistent picture
+#[allow(deprecated)]
```
Comment (epage) [inline tests/testsuite/build_script_env.rs:267]: Highlighting this for better tracking

File: tests/testsuite/build_script_env.rs [hunk 3]
```diff
@@ -198,6 +224,7 @@ fn build_script_env_verbose() {
         .run();
 }
 
+#[allow(deprecated)]
```
Comment (epage) [inline tests/testsuite/build_script_env.rs:227]: Highlighting this for better tracking

File: tests/testsuite/build_script_env.rs [hunk 4]
```diff
@@ -298,6 +326,7 @@ fn cfg_paradox() {
 /// ```
 /// CARGO_CFG_TARGET_THREAD_LOCAL=
 /// ```
+#[allow(deprecated)]
```
Comment (epage) [inline tests/testsuite/build_script_env.rs:329]: Highlighting this for better tracking

File: tests/testsuite/build_script_env.rs (hunk 2 above)
Comment (heisen-li) [reply tests/testsuite/build_script_env.rs:267]: Sorry, I don't know how to emphasize this to make it more obvious (does that mean add notes?) . Can you go into more detail or examples?

Comment (epage) [reply tests/testsuite/build_script_env.rs:267]: We had a race condition.  I was posting at the same time as you.  What you did was what I would recommend: start discussions on topics that you know reviewers need to look at.

File: tests/testsuite/build_script_env.rs (hunk 1 above)
Comment (epage) [reply tests/testsuite/build_script_env.rs:210]: Yes, for `-vv` output, we should probably still do "contains" checks as mentioned in the instructions

So
```rust
    p.cargo("check -vv")
        .with_stderr_contains("[WARNING] non-trivial mutual dependency between target-specific configuration and RUSTFLAGS")
        .with_stderr_contains("[foo 0.0.1] cfg!(bertrand)=true")
        .with_stderr_contains("[..]--cfg=bertrand[..]")
        .run();
```
could be
```rust
    p.cargo("check -vv")
        .with_stderr_contains("\
...
[WARNING] non-trivial mutual dependency between target-specific configuration and RUSTFLAGS
...")
        .with_stderr_contains("\
...
[foo 0.0.1] cfg!(bertrand)=true
...")
        .with_stderr_contains("\
...
[..]--cfg=bertrand[..]
...")
        .run();
```
`...` is a multi-line glob.  `str![]` isn't as important in these cases as we can't really update snapshots like these.

**Comment by epage** [comment]
Thanks!

@bors r+

**Pull request merged**